src/core/socket.c

```diff
@@ -141,11 +143,20 @@ void socket_free_ports(Socket *s) {
 
 static void socket_done(Unit *u) {
         Socket *s = SOCKET(u);
+        SocketPeer *p;
+        Iterator i;
 
         assert(s);
 
         socket_free_ports(s);
 
+        HASHMAP_FOREACH(p, s->peers_by_address, i) {
+                hashmap_remove(s->peers_by_address, p);
+                free(p);
+        }
```

Comment:
Yes  now I moved the clean up of the peer when the service gets stop itself. That is we remove from hashmap and free the peer memory. But we still need to clean up the hashmap memory itself.  I think there will be a memory leak . Inside the peer object if we store the hashmap pointer then we can access it after Socket object gets freed also. 

1. socket starts running let's say 3 telnet clients. Here we store the peer address that contains
the socket address.

``` 
struct SocketPeer {
        unsigned n_ref;
        Socket *s;
        int family;
        union sockaddr_union peer;
};
```

2. If we restart the socket the service object has invalid memory which is actually freed that is the pointer 's' . So we can only decrement the counter but never able remove from hash table since it's freed. And it will be there forever. How to deal with this ? I am pushing the code which is crashing when we try to remove from hashmap.